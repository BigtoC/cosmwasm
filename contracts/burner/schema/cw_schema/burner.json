{
  "contract_name": "burner",
  "contract_version": "0.0.0",
  "idl_version": "1.0.0",
  "instantiate": {
    "type": "v1",
    "root": 0,
    "definitions": [
      {
        "name": "burner_msg_InstantiateMsg",
        "description": "A placeholder where we don't take any input",
        "type": "struct",
        "properties": {}
      }
    ]
  },
  "execute": null,
  "query": null,
  "migrate": {
    "type": "v1",
    "root": 0,
    "definitions": [
      {
        "name": "burner_msg_MigrateMsg",
        "type": "struct",
        "properties": {
          "delete": {
            "description": "Optional amount of items to delete in this call.\nIf it is not provided, nothing will be deleted.\nYou can delete further items in a subsequent execute call.",
            "value": 2
          },
          "payout": {
            "description": "The address we send all remaining balance to",
            "value": 1
          }
        }
      },
      {
        "name": "String",
        "type": "string"
      },
      {
        "name": "u32",
        "type": "integer",
        "precision": 32,
        "signed": false
      }
    ]
  },
  "sudo": null,
  "responses": null
}
