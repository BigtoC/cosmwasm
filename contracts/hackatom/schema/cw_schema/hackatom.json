{
  "contract_name": "hackatom",
  "contract_version": "0.0.0",
  "idl_version": "1.0.0",
  "instantiate": {
    "type": "v1",
    "root": 0,
    "definitions": [
      {
        "name": "hackatom_msg_InstantiateMsg",
        "type": "struct",
        "properties": {
          "beneficiary": {
            "value": 1
          },
          "verifier": {
            "value": 1
          }
        }
      },
      {
        "name": "String",
        "type": "string"
      }
    ]
  },
  "execute": {
    "type": "v1",
    "root": 0,
    "definitions": [
      {
        "name": "hackatom_msg_ExecuteMsg",
        "type": "enum",
        "cases": {
          "allocate_large_memory": {
            "description": "Allocate large amounts of memory without consuming much gas",
            "type": "named",
            "properties": {
              "pages": {
                "value": 1
              }
            }
          },
          "cpu_loop": {
            "description": "Infinite loop to burn cpu cycles (only run when metering is enabled)",
            "type": "named",
            "properties": {}
          },
          "memory_loop": {
            "description": "Infinite loop reading and writing memory",
            "type": "named",
            "properties": {}
          },
          "message_loop": {
            "description": "Infinite loop sending message to itself",
            "type": "named",
            "properties": {}
          },
          "panic": {
            "description": "Trigger a panic to ensure framework handles gracefully",
            "type": "named",
            "properties": {}
          },
          "release": {
            "description": "Releasing all funds in the contract to the beneficiary. This is the only \"proper\" action of this demo contract.",
            "type": "named",
            "properties": {}
          },
          "storage_loop": {
            "description": "Infinite loop making storage calls (to test when their limit hits)",
            "type": "named",
            "properties": {}
          },
          "user_errors_in_api_calls": {
            "description": "Starting with CosmWasm 0.10, some API calls return user errors back to the contract.\nThis triggers such user errors, ensuring the transaction does not fail in the backend.",
            "type": "named",
            "properties": {}
          }
        }
      },
      {
        "name": "u32",
        "type": "integer",
        "precision": 32,
        "signed": false
      }
    ]
  },
  "query": {
    "type": "v1",
    "root": 0,
    "definitions": [
      {
        "name": "hackatom_msg_QueryMsg",
        "type": "enum",
        "cases": {
          "get_int": {
            "description": "GetInt returns a hardcoded u32 value",
            "type": "named",
            "properties": {}
          },
          "other_balance": {
            "description": "This returns cosmwasm_std::AllBalanceResponse to demo use of the querier",
            "type": "named",
            "properties": {
              "address": {
                "value": 1
              }
            }
          },
          "recurse": {
            "description": "Recurse will execute a query into itself up to depth-times and return\nEach step of the recursion may perform some extra work to test gas metering\n(`work` rounds of sha256 on contract).\nNow that we have Env, we can auto-calculate the address to recurse into",
            "type": "named",
            "properties": {
              "depth": {
                "value": 2
              },
              "work": {
                "value": 2
              }
            }
          },
          "verifier": {
            "description": "returns a human-readable representation of the verifier\nuse to ensure query path works in integration tests",
            "type": "named",
            "properties": {}
          }
        }
      },
      {
        "name": "String",
        "type": "string"
      },
      {
        "name": "u32",
        "type": "integer",
        "precision": 32,
        "signed": false
      }
    ]
  },
  "migrate": {
    "type": "v1",
    "root": 0,
    "definitions": [
      {
        "name": "hackatom_msg_MigrateMsg",
        "description": "MigrateMsg allows a privileged contract administrator to run\na migration on the contract. In this (demo) case it is just migrating\nfrom one hackatom code to the same code, but taking advantage of the\nmigration step to set a new validator.\n\nNote that the contract doesn't enforce permissions here, this is done\nby blockchain logic (in the future by blockchain governance)",
        "type": "struct",
        "properties": {
          "verifier": {
            "value": 1
          }
        }
      },
      {
        "name": "String",
        "type": "string"
      }
    ]
  },
  "sudo": {
    "type": "v1",
    "root": 0,
    "definitions": [
      {
        "name": "hackatom_msg_SudoMsg",
        "description": "SudoMsg is only exposed for internal Cosmos SDK modules to call.\nThis is showing how we can expose \"admin\" functionality than can not be called by\nexternal users or contracts, but only trusted (native/Go) code in the blockchain",
        "type": "enum",
        "cases": {
          "steal_funds": {
            "type": "named",
            "properties": {
              "amount": {
                "value": 4
              },
              "recipient": {
                "value": 1
              }
            }
          }
        }
      },
      {
        "name": "String",
        "type": "string"
      },
      {
        "name": "cosmwasm_std_coin_Coin",
        "type": "struct",
        "properties": {
          "amount": {
            "value": 3
          },
          "denom": {
            "value": 1
          }
        }
      },
      {
        "name": "cosmwasm_std_math_uint128_Uint128",
        "description": "A thin wrapper around u128 that is using strings for JSON encoding/decoding,\nsuch that the full u128 range can be used for clients that convert JSON numbers to floats,\nlike JavaScript and jq.\n\n# Examples\n\nUse `from` to create instances of this and `u128` to get the value out:\n\n```\n# use cosmwasm_std::Uint128;\nlet a = Uint128::from(123u128);\nassert_eq!(a.u128(), 123);\n\nlet b = Uint128::from(42u64);\nassert_eq!(b.u128(), 42);\n\nlet c = Uint128::from(70u32);\nassert_eq!(c.u128(), 70);\n```",
        "type": "integer",
        "precision": 128,
        "signed": false
      },
      {
        "name": "alloc::vec::Vec<cosmwasm_std::coin::Coin>",
        "type": "array",
        "items": 2
      }
    ]
  },
  "responses": {
    "get_int": {
      "type": "v1",
      "root": 0,
      "definitions": [
        {
          "name": "hackatom_msg_IntResponse",
          "type": "struct",
          "properties": {
            "int": {
              "value": 1
            }
          }
        },
        {
          "name": "u32",
          "type": "integer",
          "precision": 32,
          "signed": false
        }
      ]
    },
    "other_balance": {
      "type": "v1",
      "root": 0,
      "definitions": [
        {
          "name": "cosmwasm_std_query_bank_AllBalanceResponse",
          "type": "struct",
          "properties": {
            "amount": {
              "description": "Returns all non-zero coins held by this account.",
              "value": 4
            }
          }
        },
        {
          "name": "cosmwasm_std_coin_Coin",
          "type": "struct",
          "properties": {
            "amount": {
              "value": 3
            },
            "denom": {
              "value": 2
            }
          }
        },
        {
          "name": "String",
          "type": "string"
        },
        {
          "name": "cosmwasm_std_math_uint128_Uint128",
          "description": "A thin wrapper around u128 that is using strings for JSON encoding/decoding,\nsuch that the full u128 range can be used for clients that convert JSON numbers to floats,\nlike JavaScript and jq.\n\n# Examples\n\nUse `from` to create instances of this and `u128` to get the value out:\n\n```\n# use cosmwasm_std::Uint128;\nlet a = Uint128::from(123u128);\nassert_eq!(a.u128(), 123);\n\nlet b = Uint128::from(42u64);\nassert_eq!(b.u128(), 42);\n\nlet c = Uint128::from(70u32);\nassert_eq!(c.u128(), 70);\n```",
          "type": "integer",
          "precision": 128,
          "signed": false
        },
        {
          "name": "alloc::vec::Vec<cosmwasm_std::coin::Coin>",
          "type": "array",
          "items": 1
        }
      ]
    },
    "recurse": {
      "type": "v1",
      "root": 0,
      "definitions": [
        {
          "name": "hackatom_msg_RecurseResponse",
          "type": "struct",
          "properties": {
            "hashed": {
              "description": "hashed is the result of running sha256 \"work+1\" times on the contract's human address",
              "value": 1
            }
          }
        },
        {
          "name": "cosmwasm_std_binary_Binary",
          "description": "Binary is a wrapper around Vec<u8> to add base64 de/serialization\nwith serde. It also adds some helper methods to help encode inline.\n\nThis is only needed as serde-json-{core,wasm} has a horrible encoding for Vec<u8>.\nSee also <https://github.com/CosmWasm/cosmwasm/blob/main/docs/MESSAGE_TYPES.md>.",
          "type": "binary"
        }
      ]
    },
    "verifier": {
      "type": "v1",
      "root": 0,
      "definitions": [
        {
          "name": "hackatom_msg_VerifierResponse",
          "type": "struct",
          "properties": {
            "verifier": {
              "value": 1
            }
          }
        },
        {
          "name": "String",
          "type": "string"
        }
      ]
    }
  }
}
