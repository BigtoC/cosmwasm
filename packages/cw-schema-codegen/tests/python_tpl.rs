use cw_schema::Schemaifier;
use serde::{Deserialize, Serialize};
use std::io::Write;

#[derive(Schemaifier, Serialize, Deserialize)]
pub enum SomeEnum {
    Field1,
    Field2(u32, u32),
    Field3 { a: String, b: u32 },
    // Field4(Box<SomeEnum>),  // TODO tkulik: Do we want to support Box<T> ?
    // Field5 { a: Box<SomeEnum> },
}

#[derive(Schemaifier, Serialize, Deserialize)]
pub struct UnitStructure;

#[derive(Schemaifier, Serialize, Deserialize)]
pub struct TupleStructure(u32, String, u128);

#[derive(Schemaifier, Serialize, Deserialize)]
pub struct NamedStructure {
    a: String,
    b: u8,
    c: SomeEnum,
}

#[test]
fn simple_enum() {
    // generate the schemas for each of the above types
    let schemas = [
        cw_schema::schema_of::<SomeEnum>(),
        cw_schema::schema_of::<UnitStructure>(),
        cw_schema::schema_of::<TupleStructure>(),
        cw_schema::schema_of::<NamedStructure>(),
    ];

    // run the codegen to typescript
    for schema in schemas {
        let cw_schema::Schema::V1(schema) = schema else {
            panic!();
        };

        let output = schema
            .definitions
            .iter()
            .map(|node| {
                let mut buf = Vec::new();
                cw_schema_codegen::python::process_node(&mut buf, &schema, node).unwrap();
                String::from_utf8(buf).unwrap()
            })
            .collect::<String>();

        insta::assert_snapshot!(output);
    }
}

macro_rules! validator {
    ($typ:ty) => {{
        let a: Box<dyn FnOnce(&str) -> ()> = Box::new(|output| {
            serde_json::from_str::<$typ>(output).unwrap();
        });
        a
    }};
}

#[test]
fn assert_validity() {
    let schemas = [
        (
            "SomeEnum",
            cw_schema::schema_of::<SomeEnum>(),
            serde_json::to_string(&SomeEnum::Field1).unwrap(),
            validator!(SomeEnum),
        ),
        (
            "SomeEnum",
            cw_schema::schema_of::<SomeEnum>(),
            serde_json::to_string(&SomeEnum::Field2(10, 23)).unwrap(),
            validator!(SomeEnum),
        ),
        (
            "SomeEnum",
            cw_schema::schema_of::<SomeEnum>(),
            serde_json::to_string(&SomeEnum::Field3 {
                a: "sdf".to_string(),
                b: 12,
            })
            .unwrap(),
            validator!(SomeEnum),
        ),
        (
            "UnitStructure",
            cw_schema::schema_of::<UnitStructure>(),
            serde_json::to_string(&UnitStructure {}).unwrap(),
            validator!(UnitStructure),
        ),
        (
            "TupleStructure",
            cw_schema::schema_of::<TupleStructure>(),
            serde_json::to_string(&TupleStructure(10, "aasdf".to_string(), 2)).unwrap(),
            validator!(TupleStructure),
        ),
        (
            "NamedStructure",
            cw_schema::schema_of::<NamedStructure>(),
            serde_json::to_string(&NamedStructure {
                a: "awer".to_string(),
                b: 4,
                c: SomeEnum::Field1,
            })
            .unwrap(),
            validator!(NamedStructure),
        ),
    ];

    for (type_name, schema, example, validator) in schemas {
        let cw_schema::Schema::V1(schema) = schema else {
            unreachable!();
        };

        let schema_output = schema
            .definitions
            .iter()
            .map(|node| {
                let mut buf = Vec::new();
                cw_schema_codegen::python::process_node(&mut buf, &schema, node).unwrap();
                String::from_utf8(buf).unwrap()
            })
            .collect::<String>();

        let mut file = tempfile::NamedTempFile::with_suffix(".py").unwrap();
        file.write_all(schema_output.as_bytes()).unwrap();
        file.write(
            format!(
                "import sys; print({type_name}.model_validate_json('{example}').model_dump_json())"
            )
            .as_bytes(),
        )
        .unwrap();
        file.flush().unwrap();

        let output = std::process::Command::new("python")
            .arg(file.path())
            .output()
            .unwrap();

        assert!(
            output.status.success(),
            "stdout: {stdout}, stderr: {stderr}\n\n schema:\n {schema_output}",
            stdout = String::from_utf8_lossy(&output.stdout),
            stderr = String::from_utf8_lossy(&output.stderr),
        );

        validator(&String::from_utf8_lossy(&output.stdout))
    }
}
