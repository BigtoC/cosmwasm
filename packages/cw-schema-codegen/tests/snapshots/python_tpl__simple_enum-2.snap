---
source: packages/cw-schema-codegen/tests/python_tpl.rs
expression: output
snapshot_kind: text
---
# This code is @generated by cw-schema-codegen. Do not modify this manually.

import typing
import decimal
from pydantic import BaseModel, RootModel



class UnitStructure(RootModel):
    """
    This is a struct level documentation for unit struct
    """
    root: None
